### PR TITLE
Fix schedulers running on multiple servers. BNN-289 #resolve

### PR DESCRIPTION
**Pull request by pedrosan7os**
The proposed solution should work, but the safeguard put upon lease renewal depends on an in-memory variable, not entirely transactional. There is a remote chance for failure. A better solution could be made by changing what is stored as lease, maybe including the server id.

### CODE REVIEW

File: server/bennu-scheduler/src/main/java/org/fenixedu/bennu/scheduler/domain/SchedulerSystem.java [hunk 1]
```diff
@@ -142,13 +143,32 @@ private Boolean shouldRun() {
...
-     * 
-     * @return
+     * Set's lease time, and stores that time in memory.
+     *
+     * @return the timestamp of the created lease.
      */
     private DateTime lease() {
-        setLease(new DateTime());
-        return getLease();
+        DateTime lease = new DateTime();
```
Comment (sergiofbsilva) [inline server/bennu-scheduler/src/main/java/org/fenixedu/bennu/scheduler/domain/SchedulerSystem.java]: withMillisOfSecond(0) could be done here.

File: server/bennu-scheduler/src/main/java/org/fenixedu/bennu/scheduler/domain/SchedulerSystem.java [hunk 2]
```diff
@@ -142,13 +143,32 @@ private Boolean shouldRun() {
...
+     * be within the lease time span, or another server could race condition the ownership.
+     *
+     * Note: the timestamp is compared without milliseconds due to lack of precision from the default DB data type for DateTime.
+     *
+     * @return true if is help by current server, false otherwise.
+     */
+    private boolean isHoldingLease() {
+        final DateTime lease = getLease().withMillisOfSecond(0);
+        return lease != null && lease.equals(latestOwnedLease.withMillisOfSecond(0)) && lease.plusMinutes(getLeaseTimeMinutes())
+                .isAfterNow();
```
Comment (sergiofbsilva) [inline server/bennu-scheduler/src/main/java/org/fenixedu/bennu/scheduler/domain/SchedulerSystem.java]: withMillisOfSecond(0) is not necessary here if it is used above.

File: server/bennu-scheduler/src/main/java/org/fenixedu/bennu/scheduler/domain/SchedulerSystem.java [hunk 3]
```diff
@@ -396,7 +427,9 @@ public static void destroy() {
                 LOG.debug("interrupted consumer thread {}", consumer.getName());
                 consumer.interrupt();
             }
-            resetLease();
```
Comment (sergiofbsilva) [inline server/bennu-scheduler/src/main/java/org/fenixedu/bennu/scheduler/domain/SchedulerSystem.java]: Clear latestOwnedLease

Comment (pedrosan7os) [inline server/bennu-scheduler/src/main/java/org/fenixedu/bennu/scheduler/domain/SchedulerSystem.java]: The change was done a few lines up.

**Pull request merged**